### PR TITLE
Edit Post: Refactor and expose PluginSidebar as final API

### PR DESCRIPTION
**Pull request by gziolo**
## Description

It's time to expose Plugin APIs as final. This PR refactors all the existing Sidebar components to share the same set components and removes duplicated custom CSS styles applied to `<PluginSidebar />`. There are no changes to the public API of `<PublicSidebar />` component, other than it is going to be available under `wp.editPost.PluginSidebar`.  I think this one works great and perfectly aligns with how `<BlockControls />` or `<InspectorControls />` work with blocks `edit` method.

There were a few changes introduced internally to increase code reuse and better split responsibilities between components:
- `<Sidebar />` becomes the component which handles Slot & Fill logic. Each component is always rendered as Fill but is guarded with a conditional check which ensures that it is rendered only when it is active.
- Editor sidebars where refactored to use the same Slot & Fill behavior to align with plugins.
- `<SettingsHeader />` component was extracted with the logic that handles tabs for Block and Document settings.
- `<BlockInspectorPanel />` component was converted to `<BlockSidebar />`.
- `<PostSettings />` component was converted to `< DocumentSidebar />`.
- `<PluginSidebar />` was refactored to use already existing components rather than its own implementation. `<SidebarLayout />` was removed together with its styles.

## How Has This Been Tested?

_Example plugin to test:_

```js
const { Button, PanelBody } = wp.components;
const { dispatch } = wp.data;
const { Fragment } = wp.element;
const { __ } = wp.i18n;
const { registerPlugin } = wp.plugins;
const { PluginSidebar } = wp.editPost;
const { PluginMoreMenuItem } = wp.editPost.__experimental;

const Icon = (
	<svg width="100%" height="100%" viewBox="0 0 100 100">
		<circle cx="50" cy="50" r="40" stroke="black" strokeWidth="3" fill="red" />
	</svg>
);
const SidebarContents = () => {
	const onClose = dispatch( 'core/edit-post' ).closeGeneralSidebar;
	return (
		<PanelBody>
			<p>{ __( 'Here is the sidebar content!' ) }</p>
			<Button isPrimary={ true } onClick={ onClose }>{ __( 'Close' ) }</Button>
		</PanelBody>
	);
};
const MyPlugin = () => {
	return (
		<Fragment>
			<PluginSidebar name="my-sidebar" title="My sidebar">
				<SidebarContents />
			</PluginSidebar>
			<PluginMoreMenuItem
				name="my-menu-item"
				icon={ Icon }
				target="my-sidebar"
				type="sidebar"
			>
				{ __( 'My sidebar' ) }
			</PluginMoreMenuItem>
		</Fragment>
	);
};

registerPlugin( 'my-plugin', {
	render: MyPlugin,
} );
```

ES5 code to paste into JS console:

```js
var Button = wp.components.Button;
var PanelBody = wp.components.PanelBody;
var dispatch = wp.data.dispatch;
var Fragment = wp.element.Fragment;
var el = wp.element.createElement;
var __ = wp.i18n.__;
var registerPlugin = wp.plugins.registerPlugin;
var PluginSidebar = wp.editPost.PluginSidebar;
var PluginMoreMenuItem = wp.editPost.__experimental.PluginMoreMenuItem;

function SidebarContents() {
	var onClose = dispatch( 'core/edit-post' ).closeGeneralSidebar;
	return (
		el(
			PanelBody, {}, 
			el( 'p', {}, __( 'Here is the sidebar content!' ) ),
			el( Button, { isPrimary: true, onClick: onClose }, __( 'Close' ) ),
		)
	);
};

function MyPlugin() {
	return (
		el( Fragment, {},
			el(
				PluginSidebar,
				{
					name: 'my-sidebar',
					title: 'My sidebar'
				},
				el( SidebarContents, {} )
			),
			el(
				PluginMoreMenuItem,
				{
					name: 'my-menu-item', 
					icon: 'smiley',
					target: 'my-sidebar',
					type: 'sidebar'
				},
				__( 'My sidebar' )
			)
		)
	);
};

registerPlugin( 'my-plugin', {
	render: MyPlugin,
} );
```

#### Document
screen shot 2018-04-06 at 11 35 59

#### Block
screen shot 2018-04-06 at 11 36 09

#### Plugin
screen shot 2018-04-06 at 11 46 04

## Types of changes
Refactoring applied. There shouldn't be any noticeable changes in the UI.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code has proper inline documentation.

### CODE REVIEW

File: edit-post/components/sidebar/index.js [hunk 1]
```diff
@@ -1,46 +1,45 @@
...
+		</SidebarFill>
 	);
 };
 
-export default compose(
+const WrappedSidebar = compose(
 	withSelect( ( select ) => ( {
 		activeSidebarName: select( 'core/edit-post' ).getActiveGeneralSidebarName(),
 	} ) ),
+	ifCondition( ( { activeSidebarName, name } ) => activeSidebarName === name ),
```
Comment (gziolo) [inline edit-post/components/sidebar/index.js]: Actually, it'd trigger less re-renders when `isActive` is returned from `withSelect`.

**Comment by gziolo** [comment]
> Can you elaborate on how to test? I tried pasting the example plugin into the console, but got a "Uncaught SyntaxError: Unexpected token <" error.

I provided also ES5 version of the plugin which you can paste in the JS console.

**Comment by jasmussen** [comment]
Here's a GIF:

sidebars

Everything about this is solid. Love it. Looks and works and feels right. Ship it!

Can't wait to see the icon pinning feature 🌟

**Comment by Xyfi** [comment]
There is something wrong with the mobile view:
another_brave_new_world____ _local_wordpress_dev_ _wordpress
Both the 'x' buttons close the sidebar.

Also @aduth had the idea to reuse the sidebar API for the document and block sidebars as well, instead this PR introduces two new sidebars (`BlockSidebar` and `DocumentSidebar`).

All in all a good PR.

**Review by abotteram — CHANGES_REQUESTED** [review]
The mobile view needs to be fixed. See comment above.

**Review by atimmer — APPROVED** [review]
Having given this a quick look, that general ideas look great. I'll leave it to @Xyfi to review and test the code in more detail.

**Comment by gziolo** [comment]
> Also @aduth had the idea to reuse the sidebar API for the document and block sidebars as well, instead this PR introduces two new sidebars (`BlockSidebar` and `DocumentSidebar`). Is this something we still want?

We could do it, although it works exactly the same at the moment, but without explicitly using `registerPlugin` method. `BlockSidebar` and `DocumentSidebar` work almost exactly the same as `PluginSidebar`, but provide their own customized implementation.

**Comment by gziolo** [comment]
@Xyfi this issue with movers exists also in the `master` branch. I think it is a regression introduced in one of the recent PRs.

**Comment by gziolo** [comment]
Opened #6081 to address issues with the overlapping block controls.

**Comment by gziolo** [comment]
@Xyfi, I think I fixed all visual issues on mobile:

screen shot 2018-04-09 at 22 05 41

it aligns with `Document` settings:

screen shot 2018-04-09 at 22 06 53

**Comment by Xyfi** [comment]
> We could do it, although it works exactly the same at the moment, but without explicitly using registerPlugin method. BlockSidebar and DocumentSidebar work almost exactly the same as PluginSidebar, but provide their own customized implementation.

If it works exactly the same, doesn't that mean there is quite some duplicate code there? Also I think that [these lines](https://github.com/WordPress/gutenberg/blob/5a64d58ad270855c16d4f85186895e20f8803da7/edit-post/components/layout/index.js#L89-L91) clutter the layout file a bit.

**Review by abotteram — APPROVED** [review]

**Comment by mtias** [comment]
Great to see all the work and refinement behind these APIs!

**Pull request merged**